invokeAsync should let you call method returning Expected<T, U>

jyavenard opened the pull request:
#### 9d59d5e948151feea0c46bcf270aff3ba4aba62e
<pre>
invokeAsync should let you call method returning Expected&lt;T, U&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=265175">https://bugs.webkit.org/show_bug.cgi?id=265175</a>
<a href="https://rdar.apple.com/118671806">rdar://118671806</a>

Reviewed by Youenn Fablet.

Add createSettledPromise(Expected&amp;&amp;&gt; utility method.

Add API tests.

* Source/WTF/wtf/NativePromise.h:
(WTF::createSettledPromise):
(WTF::invokeAsync):
* Source/WTF/wtf/TypeTraits.h: Add IsExpected template.
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/271010@main">https://commits.webkit.org/271010@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66a4cdd04679e8d403356cc0463b87efe12b1483

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27072 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5688 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28308 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29280 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24753 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27525 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7559 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3083 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24610 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27334 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4509 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23239 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3944 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4030 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24236 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29916 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/23583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24712 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24647 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30231 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/26253 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4041 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2233 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28148 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5504 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33708 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6503 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4506 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7307 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4417 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -1466,23 +1466,37 @@ class NativePromiseRequest final {
     static_assert(!std::is_lvalue_reference_v<Function>, "Function object must not be passed by lvalue-ref (to avoid unplanned copies); WTFMove() the object.");
-    static_assert(IsConvertibleToNativePromise<typename RemoveSmartPointer<decltype(function())>::type>, "Function object must return Ref<NativePromise>");
     using ReturnType = typename RemoveSmartPointer<decltype(function())>::type;
+    static_assert(IsConvertibleToNativePromise<ReturnType> || IsExpected<ReturnType>::value, "Function object must return Ref<NativePromise> or Expected<T, F>");
```
You always `RemoveSmartPointer` before you test, this could misleadingly accept `NativePromise` (without `Ref`), `RefPtr<NativePromise>`, or `Ref<Expected>` (and explode with a less-friendly error below).
So you may have to be more specific, something like:
```
using ReturnType = decltype(function());
constexpr bool returnTypeIsRef = IsSmartRef<ReturnType>::value;
using ReturnTypeNoRef = typename RemoveSmartPointer<ReturnType>::type;
static_assert(returnTypeIsRef && IsConvertibleToNativePromise< ReturnTypeNoRef> || IsExpected<ReturnType>::value, "Function object must return Ref<NativePromise> or Expected");
... change the other tests&types accordingly
```

jyavenard replied:
That it could return NativePromise instead of Ref<NativePromise> is an issue indeed, however it would immediately be wrapped in a Ref<NativePromise.

Ref<Expected> wouldn't compile as it's not ref-counted anyway. but I see your point.

squelart commented on Source/WTF/wtf/TypeTraits.h:
```diff
@@ -170,4 +171,12 @@ template <typename T, typename U>
+
+template <typename T, typename E>
+struct IsExpected<Expected<T, E>> : std::true_type {
+};
```
Nit: Remove newline -> `{ };` (same as above).

squelart commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -1466,23 +1466,37 @@ class NativePromiseRequest final {
+
 // Invoke a function object (e.g., lambda) asynchronously.
 // Returns a promise that the function should eventually resolve or reject once the original promise returned by the lambda
 // is itself resolved or rejected.
```
Should this comment be updated to mention Expected?

squelart reviewed (approved):
Unofficial r+ with changes.

youennf commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -1466,23 +1466,37 @@ class NativePromiseRequest final {
+            static_cast<Ref<typename ReturnType::PromiseType>>(function())->chainTo(WTFMove(producer), { "invokeAsync proxy", nullptr });
+        });
+        return promise;
+    } else {
```
Would be good to add a code path for `void` as well to remove an unnecessary `GenericPromise::createAndResolve()` call.

jyavenard replied:
ok will do.

jyavenard committed:
invokeAsync should let you call method returning Expected<T, U>
https://bugs.webkit.org/show_bug.cgi?id=265175
rdar://118671806

Reviewed by Youenn Fablet.

Add createSettledPromise(Expected&&> utility method.

Add API tests.

* Source/WTF/wtf/NativePromise.h:
(WTF::createSettledPromise):
(WTF::invokeAsync):
* Source/WTF/wtf/TypeTraits.h: Add IsExpected template.
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/271010@main
